### PR TITLE
feat: LlamaIndex Workflows

### PR DESCRIPTION
**Pull request by noabenefraim**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [ ] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ ] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ ] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Comment by noabenefraim** [comment]
Added file to exclude.txt because smart quotes regex was failing from the check spelling linter, however the smart quotes were from the original text sources used in the NB.

File: gemini/orchestration/RAGWorkflow.png
Comment (noabenefraim) [inline gemini/orchestration/RAGWorkflow.png:1]: Looking to add this image to the repo's storage bucket. Don't seem to have access.

Once it's added I'll remove this from the PR and change the path to access in the NB.

File: .github/actions/spelling/excludes.txt [hunk 1]
```diff
@@ -103,6 +103,7 @@
 ^\Qvision/getting-started/image_generation.ipynb\E$
 ^\Qvision/getting-started/visual_captioning.ipynb\E$
 ^\Qvision/use-cases/creating_high_quality_visual_assets_with_gemini_and_imagen.ipynb\E$
+^\Qgemini/orchestration/llamaindex_workflows.ipynb\E$
```
Comment (holtskinner) [inline .github/actions/spelling/excludes.txt:106]: Can you remove this from the excludes list?

File: gemini/orchestration/RAGWorkflow.png
Comment (holtskinner) [reply gemini/orchestration/RAGWorkflow.png:1]: Added:

GS Link gs://github-repo/generative-ai/gemini/orchestration/llamaindex_workflows/RAGWorkflow.png

And the public URL https://storage.googleapis.com/github-repo/generative-ai/gemini/orchestration/llamaindex_workflows/RAGWorkflow.png

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 2]
```diff
@@ -0,0 +1,1127 @@
...
+        "+ Event-driven: This means that the workflow reacts to events as they happen, making it flexible and dynamic."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "5bdceacc7b39"
+      },
+      "source": [
+        "This notebook perform a complex Retrieval Augmented Generation (RAG) workflow using Gemini models as well as Google Firestore databases. There are two branches for this workflow:\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:70]:
```suggestion
        "This notebook perform a complex Retrieval Augmented Generation (RAG) workflow using Gemini models and Firestore databases. There are two branches for this workflow:\n",
```

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 3]
```diff
@@ -0,0 +1,1127 @@
...
+        "\n",
+        "_Branch 1_\n",
+        "+ Start Event triggered by providing a data directory to the workflow\n",
+        "+ Ingest data using the LlamaIndex SimpleDirectoryReader\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:74]:
```suggestion
        "+ Ingest data using the LlamaIndex `SimpleDirectoryReader`\n",
```

Use code font when mentioning a class or variable name in markdown

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 4]
```diff
@@ -0,0 +1,1127 @@
...
+        "+ https://docs.llamaindex.ai/en/stable/examples/workflow/citation_query_engine/"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 1,
+      "metadata": {
+        "id": "379b29265257"
+      },
+      "outputs": [
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:96]: Remove this output

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 5]
```diff
@@ -0,0 +1,1127 @@
...
+              "width": 500
+            }
+          },
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "from IPython.display import Image\n",
+        "\n",
+        "Image(\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:117]: You don't need to put this in a Python cell. Just link in the GCS image like this in a markdown cell:

```md
![RAGWorkflow](https://storage.googleapis.com/github-repo/generative-ai/gemini/orchestration/llamaindex_workflows/RAGWorkflow.png)
```

Looks like this

RAGWorkflow

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 6]
```diff
@@ -0,0 +1,1127 @@
...
+      "metadata": {
+        "id": "e8d7771a5818"
+      },
+      "source": [
+        "### Set Google Cloud project information and initialize Vertex AI SDK\n",
+        "This notebook requires the following resources:\n",
+        "+ initialized Google Cloud project\n",
+        "+ Vertex AI API enabled\n",
+        "+ existing VPC/Subnet\n",
+        "+ existing Firestore database\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:215]: Start these lines with capital letters.

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 7]
```diff
@@ -0,0 +1,1127 @@
...
+        "    SafetySetting(\n",
+        "        category=HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,\n",
+        "        threshold=HarmBlockThreshold.BLOCK_ONLY_HIGH,\n",
+        "    ),\n",
+        "]\n",
+        "embedding_model = VertexTextEmbedding(\n",
+        "    model_name=\"text-embedding-004\", credentials=credentials\n",
+        ")\n",
+        "llm = Vertex(\n",
+        "    model=\"gemini-pro\",\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:400]: Use a more specific model name. Like `gemini-1.5-pro`

File: gemini/orchestration/llamaindex_workflows.ipynb [hunk 8]
```diff
@@ -0,0 +1,1127 @@
...
+        "Settings.llm = llm"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "1b99e7e4aa0d"
+      },
+      "source": [
+        "### Define Event classes\n",
```
Comment (holtskinner) [inline gemini/orchestration/llamaindex_workflows.ipynb:417]: Why do you need to create custom Event classes? Should these be added to the LlamaIndex library directly?

**Review by holtskinner — APPROVED** [review]

**Comment by holtskinner** [comment]
Can you also add a link/description of this notebook in rag-grounding/README.md ?

File: .github/actions/spelling/excludes.txt (hunk 1 above)
Comment (noabenefraim) [reply .github/actions/spelling/excludes.txt:106]: Refer to comment above: https://github.com/GoogleCloudPlatform/generative-ai/pull/1198#issuecomment-2384143509

File: gemini/orchestration/llamaindex_workflows.ipynb (hunk 5 above)
Comment (noabenefraim) [reply gemini/orchestration/llamaindex_workflows.ipynb:117]: Done.

File: gemini/orchestration/llamaindex_workflows.ipynb (hunk 7 above)
Comment (noabenefraim) [reply gemini/orchestration/llamaindex_workflows.ipynb:400]: Done!

File: gemini/orchestration/llamaindex_workflows.ipynb (hunk 6 above)
Comment (noabenefraim) [reply gemini/orchestration/llamaindex_workflows.ipynb:215]: Done.

**Comment by noabenefraim** [comment]
> Can you also add a link/description of this notebook in rag-grounding/README.md ?

https://github.com/GoogleCloudPlatform/generative-ai/pull/1198/files#diff-402618e2bf8d99bda03742ff795ca9d12b1f9246049d2c342443404b9c215fb0 added here

File: gemini/orchestration/llamaindex_workflows.ipynb (hunk 8 above)
Comment (noabenefraim) [reply gemini/orchestration/llamaindex_workflows.ipynb:386]: LlamaIndex only provides the StartEvent and StopEvent. I believe they want them to be completely user defined.
According to LlamaIndex:

"Events are user-defined pydantic objects. You control the attributes and any other auxiliary methods. In this case, our workflow relies on a single user-defined event, the JokeEvent."

"While events are use-defined, there are two special-case events, the StartEvent and the StopEvent"

https://docs.llamaindex.ai/en/stable/module_guides/workflow/

LMK if OK to resolve @holtskinner.

Comment (holtskinner) [reply gemini/orchestration/llamaindex_workflows.ipynb:386]: Ok, sounds good!

File: .github/actions/spelling/excludes.txt (hunk 1 above)
Comment (holtskinner) [reply .github/actions/spelling/excludes.txt:106]: Ah, missed the comment. Thanks!

**Review by holtskinner — APPROVED** [review]

**Pull request merged**